### PR TITLE
Rename banned and comment_banned roles

### PR DESCRIPTION
**Pull request by jacobherrington**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Moving from "banned" terminology to "suspended."

This change helps reflect the suggestions in https://github.com/forem/internalEngineering/issues/298 in the code as a followup to https://github.com/forem/forem/pull/11581, which modified the same language in the UI.

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/11602

## Added tests?

- [ ] Yes
- [x] No, and this is why: no functionality has changed, existing tests should verify that things are working as intended
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

## [optional] Are there any post deployment tasks we need to perform?

Attached data update script, otherwise nothing.

### CODE REVIEW

**Comment by Ridhwana** [comment]
Hoping https://github.com/forem/forem/pull/12294 will fix these flakey tests

File: app/controllers/application_controller.rb
Comment (citizen428) [inline app/controllers/application_controller.rb:117]: I know you just changed the method name here, but maybe now is a good time to move this from a `RuntimeError` with a message of `"SUSPENDED"` to an actual error (see `app/errors`), like e.g. 

```ruby
# app/errors/suspended_error.rb
class SuspendedError < StandardError; end
```

File: spec/models/role_spec.rb
Comment (citizen428) [inline spec/models/role_spec.rb:11]: **nitpick (non-blocking):** These were alphabetically sorted before, but changing `banned` to `suspended` changes that.

File: lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb
Comment (citizen428) [inline lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb:6]: I don't have banned users locally, so running this is a Rails console leads to an error:

```ruby
[3] forem(main)> Role.find_by(name: "banned").update(name: "suspended")
  Role Load (0.5ms)  SELECT "roles".* FROM "roles" WHERE "roles"."name" = $1 LIMIT $2  [["name", "banned"], ["LIMIT", 1]]
NoMethodError: undefined method `update' for nil:NilClass
from (pry):3:in `<main>'
```

I then ran the data update script and checked, its status is `failed`.

We should guard against this:

```suggestion
      Role.find_by(name: "banned")&.update(name: "suspended")
```

File: lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb
Comment (citizen428) [inline lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb:7]: Same as above:

```suggestion
      Role.find_by(name: "comment_banned")&.update(name: "comment_suspended")
```

**Review by citizen428 — CHANGES_REQUESTED** [review]
Thanks for the effort, just a minor change required to make the data update script more robust.

**Comment by msarit** [comment]
I'm reviewing this now but, @jacobherrington I just wanna say you are certified #BadAss for taking this on 😅  I'll be learning a lot from this PR 🙏🏾

**Review by msarit — COMMENTED** [review]
Great work @jacobherrington! 👏🏾 

I just have a few comments but nothing blocking

File: app/models/user.rb
Comment (citizen428) [reply app/models/user.rb:368]: @msarit @jacobherrington  In general, this shouldn't be a problem because `has_role?` performs a database lookup and the data update script updates the role name there. BUT: the scripts don't run during deployment, but get scheduled to run in ~5 minutes. During this short window the role would still be called `:banned` but checks would be performed against `:suspended`. So we should indeed keep it like this for the initial deployment but can remove it after.

File: app/queries/admin/moderators_query.rb
Comment (rhymes) [inline app/queries/admin/moderators_query.rb:7]: Since `user.suspended` still looks for the `banned` role in the DB during the transition, should this query still keep `banned` and `comment_banned` as valid roles and remove them in a following PR?

File: app/services/search/query_builders/user.rb
Comment (rhymes) [inline app/services/search/query_builders/user.rb:25]: Same thing. If "legacy" users still have the `banned` role, this will show them in the search. It's probably un unlikely scenario BUT if the data update script fails or if for whatever reason takes too long, we might show banned users in the search.

**Review by rhymes — COMMENTED** [review]
One general question: is there anything in the various caches that we need to manually purge?

File: app/services/search/query_builders/user.rb
Comment (jacobherrington) [reply app/services/search/query_builders/user.rb:25]: We can leave them, then I will follow up with a PR to remove them down the road. Safety first :safety_pin:

**Comment by jacobherrington** [comment]
@rhymes I don't think so, but I'm not confident speaking definitively on that :grimacing:

File: app/queries/admin/moderators_query.rb
Comment (citizen428) [reply app/queries/admin/moderators_query.rb:7]: @rhymes See https://github.com/forem/forem/pull/12270#discussion_r562338386

**Comment by citizen428** [comment]
@jacobherrington Can we get this unblocked? I guess resolving the merge conflicts would be a first step.

File: app/models/user.rb
Comment (citizen428) [inline app/models/user.rb:366]: Unrelated, but since we're here already, should we maybe get rid of this comment, we no longer use Administrate.

File: app/models/user.rb
Comment (citizen428) [inline app/models/user.rb:367]: Maybe we should name this more idiomatically?

```suggestion
  def suspended?
```

File: app/models/user.rb
Comment (citizen428) [inline app/models/user.rb:410]: Same suggestion as above:

```suggestion
  def comment_suspended?
```

File: app/policies/comment_policy.rb
Comment (citizen428) [inline app/policies/comment_policy.rb:69]: Let's be a bit more consistent about the style we use:

```suggestion
    user.has_role?(:comment_suspended)
```

File: app/policies/application_policy.rb
Comment (citizen428) [inline app/policies/application_policy.rb:74]: Nitpick: generally it's a bit unidiomatic to use `is` in predicate methods in Ruby, `user_suspended?` would be the more common form. But that would lead to a lot of extra changes, so is probably not worth it.

```suggestion
  def user_suspended?
```

**Comment by mstruve** [comment]
Hey all, noticed this PR has been open for a bit without much activity. Any blockers I can help with?

**Comment by mstruve** [comment]
Hey @jacobherrington, this has been sitting open for a bit. Do you think you could finish it up and get it merged? If there are any blockers you are hitting please feel free to reach out so I, or someone else on the team, can help unblock you.

File: app/policies/application_policy.rb
Comment (jacobherrington) [reply app/policies/application_policy.rb]: Unfortunately, we sorta do this everywhere :laughing: I'll fix this instance in this PR and follow up with a 15-minute fix for a few other places that I've seen it!

**Comment by jacobherrington** [comment]
@rhymes Manually testing this with the cache setting enabled on my local machine seems to work as expected.

**Comment by jacobherrington** [comment]
@mstruve Thanks for the ping, I consistently forget that this PR exists. It is ready for review if you'd like to take a look at it :smile:

File: lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb
Comment (rhymes) [reply lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb:6]: this is likely going to break still

File: lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb
Comment (rhymes) [reply lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb:7]: same

File: lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb
Comment (mstruve) [inline lib/data_update_scripts/20210119131811_replace_instances_of_banned_and_comment_banned.rb:6]: Is every environment guaranteed to have these roles present? If not we should use a safety operator here and below.

**Review by rhymes — COMMENTED** [review]
Thanks for picking this up @jacobherrington!

One detail: are we testing that `user.suspended?` in the model is `true` with either role? `banned` or `suspended` ?

File: app/models/user.rb
Comment (mstruve) [inline app/models/user.rb:373]: should we put a TODO here are well to clean up the banned check once this is merged and the data update script has run?

File: app/helpers/application_helper.rb
Comment (mstruve) [inline app/helpers/application_helper.rb:293]: It might be helpful to reference here what will allow this to be removed. For example, once this PR has been merged and deployed. That way if we do accidentally forget about it when someone comes along they have the necessary context to know when and under what circumstances this is safe to remove.

**Review by mstruve — COMMENTED** [review]
Thanks so much for closing the loop on this PR, I'm excited to get this out the doo. Probably needs someone to go through and give it a good test but I left a couple of small comments.

**Comment by jacobherrington** [comment]
@mstruve I clarified the comments to indicate when we can removed references to banned and comment_banned.

@rhymes it's a little hard to test because the banned role doesn't "exist" anymore - should we force the user model to behave a way it won't in production in a test?

**Review by rhymes — APPROVED** [review]
@jacobherrington got it, this PR assumes that the update script has been run successfully. 

In theory you could add `banned` as a role in the test and check the result of `.suspended?` is still `true` but since the role is going to leave at most a few days (non DEV Forems aren't updated automatically), we should be confident enough nobody will change the code in the meantime :) 

Good job!

**Review by citizen428 — APPROVED** [review]

**Pull request merged**
